infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java

```diff
@@ -114,127 +102,152 @@ private void recordRunnable(Runnable fsyncRunnable, Optional<Timer> fsyncTimer)
+                        for (BatchWriterOperation op : res) {
+                            if (!op.getFutureResult().isCompletedExceptionally()
+                                    && !op.getFutureResult().isCancelled()) {
+                                op.getFutureResult().complete(op.getResultValue());
                             }
                         }
+
                         res.clear();
-                        processed = 0;
+                        numProcessed = 0;
                     }
                 }
 
-                if (currOp == null) {
+                if (currentOp == null) {
                     lastOp = null;
-                } else if (currOp == BatchWriterOperation.SHUTDOWN) {
-                    log.warn("Shutting down the write processor");
+                } else if (currentOp == BatchWriterOperation.SHUTDOWN) {
+                    log.warn("batchWriteProcessor: shutting down the write processor");
                     streamLog.sync(true);
                     break;
-                } else if (streamLog.quotaExceeded() && currOp.getMsg().getPriorityLevel() != PriorityLevel.HIGH) {
-                    currOp.getFutureResult().completeExceptionally(
-                            new QuotaExceededException("Quota of "
-                                    + streamLog.quotaLimitInBytes() + " bytes"));
-                    log.warn("batchprocessor: quota exceeded, dropping msg {}", currOp.getMsg());
-                } else if (currOp.getType() == Type.SEAL && currOp.getMsg().getEpoch() >= sealEpoch) {
-                    log.info("batchWriteProcessor: updating from {} to {}", sealEpoch, currOp.getMsg().getEpoch());
-                    sealEpoch = currOp.getMsg().getEpoch();
-                    res.add(currOp);
-                    processed++;
-                    lastOp = currOp;
-                } else if (currOp.getMsg().getEpoch() != sealEpoch) {
-                    log.warn("batchWriteProcessor: wrong epoch on {} msg, seal epoch is {}, and msg epoch is {}",
-                            currOp.getType(), sealEpoch, currOp.getMsg().getEpoch());
-                    currOp.getFutureResult().completeExceptionally(new WrongEpochException(sealEpoch));
-                    res.add(currOp);
-                    processed++;
-                    lastOp = currOp;
+                } else if (streamLog.quotaExceeded() &&
+                        (currentOp.getRequest().getHeader().getPriority() != PriorityLevel.HIGH)) {
+                    currentOp.getFutureResult().completeExceptionally(
+                            new QuotaExceededException("Quota of " + streamLog.quotaLimitInBytes() + " bytes"));
+
+                    log.warn("batchWriteProcessor: quota exceeded, dropping request {}",
+                            TextFormat.shortDebugString(currentOp.getRequest()));
+                } else if (currentOp.getType() == BatchWriterOperation.Type.SEAL &&
```

Comment:
I think we can remove `BatchWriterOperation.Type` since all the message types map to an operation in this processor, except for shutdown, which can be encoded as a variable instead of a message in the work queue.

Reply:
Then you can check and map the message type directly. For example, `if (currentOp.getRequest().getPayload().hasSealRequest())`